### PR TITLE
documentation #546: added swagger documentation for 5 Group endpoints:

### LINKED ISSUE #546: Add Swagger documentation to Group endpoints (2)
Add documentation to Swagger **only** for these endpoints:

image

### PR DESCRIPTION
**Pull request by vlad-zhadan**
-/students
-/captain
-/disciplineTeachers
-/discipline
-/addEmails

closes #546

### CODE REVIEW

File: src/v2/api/controllers/GroupController.ts [hunk 1]
```diff
@@ -122,54 +141,118 @@ export class GroupController {
...
 
+  @ApiBearerAuth()
+  @ApiOkResponse({
+    type: ShortUsersResponse,
+  })
+  @ApiBadRequestResponse({
+    description: `
+      InvalidEntityIdException: Group with such id is not found
+      InvalidBodyException: Email can not be empty
+      InvalidBodyException: One of emails is not an email
```
Comment (SharpDevOps10) [inline src/v2/api/controllers/GroupController.ts:245]: Change this description to : `The email is not a valid email address`

File: src/v2/api/dtos/EmailDTO.ts [hunk 2]
```diff
@@ -1,13 +1,23 @@
...
 import { validationOptionsMsg } from '../../utils/GLOBALS';
+import { ApiProperty } from '@nestjs/swagger';
 
 export class EmailDTO {
-    @ArrayMinSize(1, validationOptionsMsg('Email can not be empty'))
-    @ArrayMaxSize(50, validationOptionsMsg('Email can not be empty'))
-    @ArrayUnique({}, validationOptionsMsg('There are duplicate emails'))
-    @IsEmail({}, {
+  @ApiProperty()
+  @ArrayMinSize(1, validationOptionsMsg('Email can not be empty'))
```
Comment (SharpDevOps10) [inline src/v2/api/dtos/EmailDTO.ts:12]: Change it to : `Email cannot be empty`

File: src/v2/api/dtos/EmailDTO.ts [hunk 3]
```diff
@@ -1,13 +1,23 @@
...
+  @ArrayMaxSize(50, validationOptionsMsg('Email can not be empty'))
```
Comment (SharpDevOps10) [inline src/v2/api/dtos/EmailDTO.ts:13]: The same as above

File: src/v2/api/dtos/EmailDTO.ts [hunk 4]
```diff
@@ -1,13 +1,23 @@
...
+  @ArrayUnique({}, validationOptionsMsg('There are duplicate emails'))
+  @IsEmail(
+    {},
+    {
       each: true,
       ...validationOptionsMsg('One of emails is not an email'),
```
Comment (SharpDevOps10) [inline src/v2/api/dtos/EmailDTO.ts:19]: Change this description to : `The email is not a valid email address`

**Review by SharpDevOps10 — CHANGES_REQUESTED** [review]

File: src/v2/api/controllers/GroupController.ts (hunk 1 above)
Comment (SharpDevOps10) [reply src/v2/api/controllers/GroupController.ts:245]: Also on line 244, change the description to: `Email cannot be empty`

File: src/v2/api/controllers/GroupController.ts [hunk 5]
```diff
@@ -122,54 +141,118 @@ export class GroupController {
...
     return this.groupService.getDisciplineTeachers(groupId);
   }
 
+  @ApiBearerAuth()
+  @ApiOkResponse({
+    type: [ShortDisciplineResponse],
+  })
+  @ApiBadRequestResponse({
+    description: `
+      InvalidEntityIdException: Group with such id is not found
```
[the remark this answers is not in the document]
Comment (PavloVlasiuk) [reply src/v2/api/controllers/GroupController.ts:222]: The exception must be "InvalidGroupIdException"

File: src/v2/api/controllers/GroupController.ts (hunk 1 above)
Comment (PavloVlasiuk) [reply src/v2/api/controllers/GroupController.ts:245]: "InvalidEntityIdException" must be "InvalidGroupIdException"

**Review by PavloVlasiuk — CHANGES_REQUESTED** [review]

File: src/v2/api/controllers/UserController.ts [hunk 6]
```diff
@@ -58,15 +59,14 @@ export class UserController {
   constructor (
     private userService: UserService,
     private userMapper: UserMapper,
-    private studentMapper: StudentMapper,
-  ) {
-  }
+    private studentMapper: StudentMapper
```
Comment (SharpDevOps10) [inline src/v2/api/controllers/UserController.ts:62]: Add missing comma

File: src/v2/api/controllers/UserController.ts [hunk 7]
```diff
@@ -188,7 +182,7 @@ export class UserController {
   @Post('/:userId/contacts')
   createContact (
     @Param('userId', UserByIdPipe) userId: string,
-    @Body() body: CreateContactDTO,
+    @Body() body: CreateContactDTO
```
Comment (SharpDevOps10) [inline src/v2/api/controllers/UserController.ts:185]: Add missing comma

File: src/v2/api/controllers/UserController.ts [hunk 8]
```diff
@@ -282,13 +268,12 @@ export class UserController {
   @UseInterceptors(FileInterceptor('avatar'))
   async uploadAvatar (
     @Param('userId', UserByIdPipe) userId: string,
-    @UploadedFile(AvatarValidationPipe) file: Express.Multer.File,
+    @UploadedFile(AvatarValidationPipe) file: Express.Multer.File
```
Comment (SharpDevOps10) [inline src/v2/api/controllers/UserController.ts:271]: Add missing comma

File: src/v2/api/controllers/UserController.ts [hunk 9]
```diff
@@ -303,8 +288,8 @@ export class UserController {
                   You do not have permission to perform this action`,
   })
   async getRemainingSelective (
-      @Param('userId', UserByIdPipe) userId: string,
-      @Query() body: RemainingSelectiveDTO,
+    @Param('userId', UserByIdPipe) userId: string,
+    @Query() body: RemainingSelectiveDTO
```
Comment (SharpDevOps10) [inline src/v2/api/controllers/UserController.ts:292]: Add missing comma

File: src/v2/api/controllers/UserController.ts [hunk 10]
```diff
@@ -329,21 +314,20 @@ export class UserController {
   })
   async attachSelectiveDisciplines (
     @Param('userId', UserByIdPipe, StudentPipe) userId: string,
-    @Body(SelectiveDisciplinesPipe) body: AttachSelectiveDisciplinesDTO,
+    @Body(SelectiveDisciplinesPipe) body: AttachSelectiveDisciplinesDTO
```
Comment (SharpDevOps10) [inline src/v2/api/controllers/UserController.ts:317]: Add missing comma

File: src/v2/api/controllers/UserController.ts [hunk 11]
```diff
@@ -352,7 +336,7 @@ export class UserController {
   })
   async changeGroup (
     @Param('userId', UserByIdPipe, StudentPipe) userId: string,
-    @Param('groupId', GroupByIdPipe) groupId: string,
+    @Param('groupId', GroupByIdPipe) groupId: string
```
Comment (SharpDevOps10) [inline src/v2/api/controllers/UserController.ts:339]: Add missing comma

File: src/v2/api/responses/TeacherResponse.ts [hunk 12]
```diff
@@ -1,27 +1,36 @@
...
+
   @ApiPropertyOptional()
     avatar?: string;
 
   @ApiProperty()
     rating: number;
 }
 
+export class ExtendTeacherResponse extends TeacherResponse {
+@ApiProperty()
```
Comment (SharpDevOps10) [inline src/v2/api/responses/TeacherResponse.ts:28]: There must be 2 spaces

File: src/v2/api/responses/TeacherResponse.ts [hunk 13]
```diff
@@ -1,27 +1,36 @@
...
+  disciplineTeacherId: string;
```
Comment (SharpDevOps10) [inline src/v2/api/responses/TeacherResponse.ts:29]: Thus, there must be 4 spaces

File: src/v2/mappers/DisciplineMapper.ts [hunk 14]
```diff
@@ -66,11 +69,15 @@ export class DisciplineMapper {
     return periods;
   }
 
-  getSelectiveWithAmount (disciplines: DbDiscipline[], amounts: SelectiveAmount[]) {
+  getSelectiveWithAmount (
+    disciplines: DbDiscipline[],
+    amounts: SelectiveAmount[]
```
Comment (SharpDevOps10) [inline src/v2/mappers/DisciplineMapper.ts:74]: Add missing comma

**Review by SharpDevOps10 — CHANGES_REQUESTED** [review]

File: src/v2/api/controllers/GroupController.ts [hunk 15]
```diff
@@ -122,51 +141,119 @@ export class GroupController {
...
+  async deleteGroup (@Param('groupId', GroupByIdPipe) groupId: string) {
     return this.groupService.deleteGroup(groupId);
   }
 
+  @ApiBearerAuth()
+  @ApiOkResponse({
+    type: GroupStudentsResponse,
+  })
+  @ApiBadRequestResponse({
+    description: `\n
```
[the remark this answers is not in the document]
Comment (vlad-zhadan) [reply src/v2/api/controllers/GroupController.ts:153]: I followed the link to adjust my code style: https://gist.github.com/Hahhhayam/43197ad702006f475eb8d97fb8c08814

File: src/v2/api/controllers/GroupController.ts [hunk 16]
```diff
@@ -141,6 +172,25 @@ export class GroupController {
     return { students };
   }
 
+  @ApiBearerAuth()
+  @ApiOkResponse({
+    type: CaptainResponse,
+  })
+  @ApiBadRequestResponse({
+    description: `\n
+  InvalidGroupIdException:
```
Comment (SharpDevOps10) [inline src/v2/api/controllers/GroupController.ts:181]: Add 2 spaces

File: src/v2/api/controllers/GroupController.ts [hunk 17]
```diff
@@ -141,6 +172,25 @@ export class GroupController {
...
   }
 
+  @ApiBearerAuth()
+  @ApiOkResponse({
+    type: CaptainResponse,
+  })
+  @ApiBadRequestResponse({
+    description: `\n
+  InvalidGroupIdException: 
+    Group with such id is not found`,
```
Comment (SharpDevOps10) [inline src/v2/api/controllers/GroupController.ts:182]: And here. Add 2 spaces

File: src/v2/api/controllers/UserController.ts [hunk 18]
```diff
@@ -171,7 +170,7 @@ export class UserController {
   @Patch('/:userId')
   async updateUser (
     @Param('userId', UserByIdPipe) userId: string,
-    @Body() body: UpdateUserDTO,
+    @Body() body: UpdateUserDTO
```
Comment (SharpDevOps10) [inline src/v2/api/controllers/UserController.ts:173]: Add missing comma

File: src/v2/mappers/DisciplineMapper.ts (hunk 14 above)
Comment (SharpDevOps10) [inline src/v2/mappers/DisciplineMapper.ts:74]: Comma?

**Review by SharpDevOps10 — CHANGES_REQUESTED** [review]

File: src/v2/api/responses/DisciplineResponse.ts [hunk 19]
```diff
@@ -1,24 +1,40 @@
...
+    type: SubjectResponse,
+  })
+    subject: SubjectResponse;
+}
+
+export class ShortDisciplinesResponse {
+  @ApiProperty({
+    type: [ShortDisciplineResponse],
+  })
+    disciplines: SubjectResponse[];
```
Comment (just-hoshinon) [inline src/v2/api/responses/DisciplineResponse.ts:39]: should be ShortDisciplinesResponse

File: src/v2/api/responses/DisciplineResponse.ts [hunk 20]
```diff
@@ -1,24 +1,40 @@
...
-  
+    groupId: string;
+
   @ApiProperty()
     description: string;
-}
\ No newline at end of file
+}
+
+export class ShortDisciplineResponse extends BaseDisciplineResponse {
```
Comment (just-hoshinon) [inline src/v2/api/responses/DisciplineResponse.ts:28]: check whether this class should contain group object or not

**Review by just-hoshinon — CHANGES_REQUESTED** [review]

**Review by PavloVlasiuk — APPROVED** [review]

**Review by SharpDevOps10 — APPROVED** [review]

**Review by basarabst — APPROVED** [review]

**Review by just-hoshinon — APPROVED** [review]

**Pull request merged**